Detectors/MUON/MCH/CTF/include/MCHCTF/CTFCoder.h

```diff
@@ -0,0 +1,156 @@
+  auto szIni = sizeof(CTFHeader) + helper.getSize() * 2. / 3; // will be autoexpanded if needed
+  buff.resize(szIni);
+
+  auto ec = CTF::create(buff);
+  using ECB = CTF::base;
+
+  ec->setHeader(helper.createHeader());
+  ec->getANSHeader().majorVersion = 0;
+  ec->getANSHeader().minorVersion = 1;
+  // at every encoding the buffer might be autoexpanded, so we don't work with fixed pointer ec
+#define ENCODEMCH(beg, end, slot, bits) CTF::get(buff.data())->encode(beg, end, int(slot), bits, optField[int(slot)], &buff, mCoders[int(slot)].get());
+  // clang-format off
+  ENCODEMCH(helper.begin_bcIncROF(),    helper.end_bcIncROF(),     CTF::BLC_bcIncROF,    0);
+  ENCODEMCH(helper.begin_orbitIncROF(), helper.end_orbitIncROF(),  CTF::BLC_orbitIncROF, 0);
+  ENCODEMCH(helper.begin_nDigitsROF(),  helper.end_nDigitsROF(),   CTF::BLC_nDigitsROF,  0);
+
+  ENCODEMCH(helper.begin_tfTime(),      helper.end_tfTime(),       CTF::BLC_tfTime,      0);
+  ENCODEMCH(helper.begin_nSamples(),    helper.end_nSamples(),     CTF::BLC_nSamples,    0);
+  ENCODEMCH(helper.begin_detID(),       helper.end_detID(),        CTF::BLC_detID,       0);
+  ENCODEMCH(helper.begin_padID(),       helper.end_padID(),        CTF::BLC_padID,       0);
+  ENCODEMCH(helper.begin_ADC()  ,       helper.end_ADC(),          CTF::BLC_ADC,         0);
+  // clang-format on
+  CTF::get(buff.data())->print(getPrefix());
+}
+
+/// decode entropy-encoded clusters to standard compact clusters
+template <typename VROF, typename VCOL>
+void CTFCoder::decode(const CTF::base& ec, VROF& rofVec, VCOL& digVec)
+{
+  auto header = ec.getHeader();
+  ec.print(getPrefix());
+  std::vector<uint16_t> bcInc, nDigits, nSamples;
+  std::vector<uint32_t> orbitInc, ADC;
+  std::vector<int32_t> tfTime;
+  std::vector<int16_t> detID, padID;
+
+#define DECODEMCH(part, slot) ec.decode(part, int(slot), mCoders[int(slot)].get())
+  // clang-format off
+  DECODEMCH(bcInc,       CTF::BLC_bcIncROF);
+  DECODEMCH(orbitInc,    CTF::BLC_orbitIncROF);
+  DECODEMCH(nDigits,     CTF::BLC_nDigitsROF);
+
+  DECODEMCH(tfTime,      CTF::BLC_tfTime);
+  DECODEMCH(nSamples,    CTF::BLC_nSamples);
+  DECODEMCH(detID,       CTF::BLC_detID);
+  DECODEMCH(padID,       CTF::BLC_padID);
+  DECODEMCH(ADC,         CTF::BLC_ADC);
+  // clang-format on
+  //
+  rofVec.clear();
+  digVec.clear();
+  rofVec.reserve(header.nROFs);
+  digVec.reserve(header.nDigits);
+
+  uint32_t firstEntry = 0, rofCount = 0, digCount = 0;
+  o2::InteractionRecord ir(header.firstBC, header.firstOrbit);
+  constexpr uint16_t SMask = 0x1 << 12, NSMask = SMask - 1;
+  for (uint32_t irof = 0; irof < header.nROFs; irof++) {
+    // restore ROFRecord
+    if (orbitInc[irof]) {  // non-0 increment => new orbit
+      ir.bc = bcInc[irof]; // bcInc has absolute meaning
+      ir.orbit += orbitInc[irof];
+    } else {
+      ir.bc += bcInc[irof];
+    }
+
+    firstEntry = digVec.size();
+    for (uint8_t ic = 0; ic < nDigits[irof]; ic++) {
```

Comment:
@shahor02 (I know this one is closed but I'm using it to get a link to the code ;-) )

While trying to get a circular test of mch ctf (https://alice.its.cern.ch/jira/browse/MRRTF-111?focusedCommentId=263799&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-263799) I got a crash in the decoding phase, that took me a while to debug.

Turns out the issue is the above line where it is assumed that the number of digits per ROF fits within 8 bits, which is not the case. I guess that's just a trivial bug (fixed in my version and it seems to work fine) but I wanted to cross-check with you that there's nowhere an underlying assumption about the max. number of items per ROF that I may have missed.

Reply:
@aphecetche right, this is a typo, sorry for that. Actually, the iterator and encoding assumes that the number of digits fits to uint16_t, see https://github.com/shahor02/AliceO2/blob/52d9dc908deb75b660d0999bf4699af49414edc9/Detectors/MUON/MCH/CTF/include/MCHCTF/CTFHelper.h#L125 , https://github.com/shahor02/AliceO2/blob/52d9dc908deb75b660d0999bf4699af49414edc9/Detectors/MUON/MCH/CTF/include/MCHCTF/CTFCoder.h#L108 and https://github.com/shahor02/AliceO2/blob/52d9dc908deb75b660d0999bf4699af49414edc9/Detectors/MUON/MCH/CTF/src/CTFCoder.cxx#L63
If you think this is not enough, just change nDigits to uint32_t on these lines.

BTW, don't remember if I've asked you: are your digits sorted in ``detID``? If yes, then you can gain little bit in compression by storing the difference of detID for all digits except the 1st one in ROF, see how I've implemented this for TRD (requires bit more complex iterator):
https://github.com/AliceO2Group/AliceO2/blob/366eaa6a631ad18342fc91cc3d31235ab18b90aa/Detectors/TRD/reconstruction/include/TRDReconstruction/CTFHelper.h#L216-L231

https://github.com/AliceO2Group/AliceO2/blob/366eaa6a631ad18342fc91cc3d31235ab18b90aa/Detectors/TRD/reconstruction/include/TRDReconstruction/CTFHelper.h#L297-L298

https://github.com/AliceO2Group/AliceO2/blob/366eaa6a631ad18342fc91cc3d31235ab18b90aa/Detectors/TRD/reconstruction/include/TRDReconstruction/CTFHelper.h#L43-L45